### ISSUE TITLE
2.9.3: test suite fails with libxml 2.12.x

### ISSUE THREAD

**Issue by kloczek** [open-mpi/hwloc#639]
Looks like when hwloc 2.9.3 is build and tested against libxml +2.12.0 test suite one unit fails:
```console
+ cd hwloc-2.9.3
+ /usr/bin/make -O -j48 V=1 VERBOSE=1 check
Making check in include
make[1]: Nothing to be done for 'check'.
Making check in hwloc
/usr/bin/make  hwloc_fake.la
make[2]: Entering directory '/home/tkloczko/rpmbuild/BUILD/hwloc-2.9.3/hwloc'
/bin/sh ../libtool  --tag=CC   --mode=compile /usr/bin/gcc -DHAVE_CONFIG_H -I. -I../include/private/autogen -I../include/hwloc/autogen  -I"/home/tkloczko/rpmbuild/BUILD/hwloc-2.9.3"/include -I"/home/tkloczko/rpmbuild/BUILD/hwloc-2.9.3"/include -DHWLOC_INSIDE_LIBHWLOC -DHWLOC_PLUGINS_PATH="\"/usr/lib64/modules/hwloc\"" -DRUNSTATEDIR="\"/run\""  -fvisibility=hidden -O2 -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-U_FORTIFY_SOURCE,-D_FORTIFY_SOURCE=3 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -fno-omit-frame-pointer -mno-omit-leaf-frame-pointer -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -fdata-sections -ffunction-sections -flto=auto -flto-partition=none -Os -c -o topology-fake.lo topology-fake.c
libtool: compile:  /usr/bin/gcc -DHAVE_CONFIG_H -I. -I../include/private/autogen -I../include/hwloc/autogen -I/home/tkloczko/rpmbuild/BUILD/hwloc-2.9.3/include -I/home/tkloczko/rpmbuild/BUILD/hwloc-2.9.3/include -DHWLOC_INSIDE_LIBHWLOC -DHWLOC_PLUGINS_PATH=\"/usr/lib64/modules/hwloc\" -DRUNSTATEDIR=\"/run\" -fvisibility=hidden -O2 -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-U_FORTIFY_SOURCE,-D_FORTIFY_SOURCE=3 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -fno-omit-frame-pointer -mno-omit-leaf-frame-pointer -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -fdata-sections -ffunction-sections -flto=auto -flto-partition=none -Os -c topology-fake.c  -fPIC -DPIC -o .libs/topology-fake.o
make[2]: Leaving directory '/home/tkloczko/rpmbuild/BUILD/hwloc-2.9.3/hwloc'
make[2]: Entering directory '/home/tkloczko/rpmbuild/BUILD/hwloc-2.9.3/hwloc'
/bin/sh ../libtool  --tag=CC   --mode=link /usr/bin/gcc -fvisibility=hidden -O2 -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-U_FORTIFY_SOURCE,-D_FORTIFY_SOURCE=3 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -fno-omit-frame-pointer -mno-omit-leaf-frame-pointer -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -fdata-sections -ffunction-sections -flto=auto -flto-partition=none -Os -module -avoid-version   -rpath /nowhere  -Wl,--gc-sections -Wl,--as-needed -flto=auto -flto-partition=none -fuse-linker-plugin -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -Wl,--build-id=sha1 -o hwloc_fake.la  topology-fake.lo
libtool: link: /usr/bin/gcc -shared  -fPIC -DPIC  .libs/topology-fake.o   -O2 -g -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -mno-omit-leaf-frame-pointer -m64 -mtune=generic -flto=auto -flto-partition=none -Os -Wl,--gc-sections -Wl,--as-needed -flto=auto -flto-partition=none -fuse-linker-plugin -Wl,-z -Wl,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -Wl,--build-id=sha1   -Wl,-soname -Wl,hwloc_fake.so -o .libs/hwloc_fake.so
libtool: link: ( cd ".libs" && rm -f "hwloc_fake.la" && ln -s "../hwloc_fake.la" "hwloc_fake.la" )
make[2]: Leaving directory '/home/tkloczko/rpmbuild/BUILD/hwloc-2.9.3/hwloc'
Making check in utils
Making check in hwloc
Making check in .
/usr/bin/make  check-TESTS
make[5]: Entering directory '/home/tkloczko/rpmbuild/BUILD/hwloc-2.9.3/utils/hwloc'
FAIL: test-hwloc-diffpatch.sh
make[5]: Leaving directory '/home/tkloczko/rpmbuild/BUILD/hwloc-2.9.3/utils/hwloc'
make[5]: Entering directory '/home/tkloczko/rpmbuild/BUILD/hwloc-2.9.3/utils/hwloc'
PASS: test-parsing-flags.sh
make[5]: Leaving directory '/home/tkloczko/rpmbuild/BUILD/hwloc-2.9.3/utils/hwloc'
make[5]: Entering directory '/home/tkloczko/rpmbuild/BUILD/hwloc-2.9.3/utils/hwloc'
PASS: test-fake-plugin.sh
make[5]: Leaving directory '/home/tkloczko/rpmbuild/BUILD/hwloc-2.9.3/utils/hwloc'
make[5]: Entering directory '/home/tkloczko/rpmbuild/BUILD/hwloc-2.9.3/utils/hwloc'
PASS: test-build-custom-topology.sh
make[5]: Leaving directory '/home/tkloczko/rpmbuild/BUILD/hwloc-2.9.3/utils/hwloc'
make[5]: Entering directory '/home/tkloczko/rpmbuild/BUILD/hwloc-2.9.3/utils/hwloc'
PASS: test-hwloc-distrib.sh
make[5]: Leaving directory '/home/tkloczko/rpmbuild/BUILD/hwloc-2.9.3/utils/hwloc'
make[5]: Entering directory '/home/tkloczko/rpmbuild/BUILD/hwloc-2.9.3/utils/hwloc'
PASS: test-hwloc-info.sh
make[5]: Leaving directory '/home/tkloczko/rpmbuild/BUILD/hwloc-2.9.3/utils/hwloc'
make[5]: Entering directory '/home/tkloczko/rpmbuild/BUILD/hwloc-2.9.3/utils/hwloc'
PASS: test-hwloc-annotate.sh
make[5]: Leaving directory '/home/tkloczko/rpmbuild/BUILD/hwloc-2.9.3/utils/hwloc'
make[5]: Entering directory '/home/tkloczko/rpmbuild/BUILD/hwloc-2.9.3/utils/hwloc'
PASS: test-hwloc-calc.sh
make[5]: Leaving directory '/home/tkloczko/rpmbuild/BUILD/hwloc-2.9.3/utils/hwloc'
make[5]: Entering directory '/home/tkloczko/rpmbuild/BUILD/hwloc-2.9.3/utils/hwloc'
PASS: test-hwloc-compress-dir.sh
make[5]: Leaving directory '/home/tkloczko/rpmbuild/BUILD/hwloc-2.9.3/utils/hwloc'
make[5]: Entering directory '/home/tkloczko/rpmbuild/BUILD/hwloc-2.9.3/utils/hwloc'
=============================================
   hwloc 2.9.3: utils/hwloc/test-suite.log
=============================================

# TOTAL: 9
# PASS:  8
# SKIP:  0
# XFAIL: 0
# FAIL:  1
# XPASS: 0
# ERROR: 0

.. contents:: :depth: 2

FAIL: test-hwloc-diffpatch.sh
=============================

Found 11 differences, exporting to stdout
:17: parser error : Extra content at the end of the document

^
Failed to load XML topology diff -
FAIL test-hwloc-diffpatch.sh (exit status: 1)

============================================================================
Testsuite summary for hwloc 2.9.3
============================================================================
# TOTAL: 9
# PASS:  8
# SKIP:  0
# XFAIL: 0
# FAIL:  1
# XPASS: 0
# ERROR: 0
============================================================================
See utils/hwloc/test-suite.log
Please report to https://github.com/open-mpi/hwloc/issues
============================================================================
make[5]: *** [Makefile:1381: test-suite.log] Error 1
make[5]: Leaving directory '/home/tkloczko/rpmbuild/BUILD/hwloc-2.9.3/utils/hwloc'
make[4]: *** [Makefile:1489: check-TESTS] Error 2
make[3]: *** [Makefile:1642: check-am] Error 2
make[2]: *** [Makefile:1273: check-recursive] Error 1
make[1]: *** [Makefile:487: check-recursive] Error 1
make: *** [Makefile:664: check-recursive] Error 1
```

**Comment by bgoglin** [comment]
Hello. Modify utils/hwloc/test-hwloc-diffpatch.sh in your build by adding "set -x" near "set -e" at the end. Then also add "cat $diffoutput" just before "cat $diffoutput | $patch refname - $output1". Run it again and send the full output.

**Comment by bgoglin** [comment]
I managed to reproduce. xmlReadMemory() does not accept inputs that were accepted in the past (looks like the ending \0 should not be counted in the buffer size). I am trying to find out if this is documented somewhere and expected, I'll open an issue upstream if I can't find a clear answer.

**Comment by kloczek** [comment]
Thx to let me know 👍

**Comment by bgoglin** [comment]
This should fix the bug:
```
--- a/hwloc/topology-xml-libxml.c
+++ b/hwloc/topology-xml-libxml.c
@@ -272,7 +272,7 @@ hwloc_libxml_import_diff(struct hwloc__xml_import_state_s *state, const char *xm
   if (xmlpath)
     doc = xmlReadFile(xmlpath, NULL, XML_PARSE_NOBLANKS);
   else if (xmlbuffer)
-    doc = xmlReadMemory(xmlbuffer, xmlbuflen, "", NULL, XML_PARSE_NOBLANKS);
+    doc = xmlReadMemory(xmlbuffer, xmlbuflen-1 /* no \0 */, "", NULL, XML_PARSE_NOBLANKS);
 
   if (!doc) {
     if (!errno)
@@ -368,7 +368,7 @@ hwloc_libxml_backend_init(struct hwloc_xml_backend_data_s *bdata,
   if (xmlpath)
     doc = xmlReadFile(xmlpath, NULL, XML_PARSE_NOBLANKS);
   else if (xmlbuffer)
-    doc = xmlReadMemory(xmlbuffer, xmlbuflen, "", NULL, XML_PARSE_NOBLANKS);
+    doc = xmlReadMemory(xmlbuffer, xmlbuflen-1, "", NULL, XML_PARSE_NOBLANKS);
 
   if (!doc) {
     if (!errno)
```
I asked upstream to confirm that the change is intented https://gitlab.gnome.org/GNOME/libxml2/-/issues/638

**Comment by bgoglin** [comment]
FYI, this is fixed in the 2.10.0 release that I am publishing right now. There's no plan to release a 2.9.4.

**Comment by kloczek** [comment]
> FYI, this is fixed in the 2.10.0 release that I am publishing right now. There's no plan to release a 2.9.4.

OK will test that ASP.

BTW ..  is it possible to start making github releases?🤔
On create github release entry is created email notification to those whom have set in your repo the web UI Watch->Releases.
gh release can contain additional comments (li changelog) or additional assets like release tar balls (by default it contains only assets from git tag) however all those part are not obligatory.
In simplest variant gh release can be empty because subiekt of the sent email contains git tag name.

I'm asking because my automation process uses those email notifications by trying to make preliminary automated upgrades of building packages, which allows saving some time on maintaining packaging procedures.
Probably other people may be interested to be instantly informed about release new version as well.

Documentation and examples of generate gh releases:
https://docs.github.com/en/repositories/releasing-projects-on-github/managing-releases-in-a-repository
https://cli.github.com/manual/gh_release_upload/
https://github.com/jbms/sphinx-immaterial/pull/282
https://github.com/marketplace/actions/github-release
https://pgjones.dev/blog/trusted-plublishing-2023/
https://github.com/jbms/sphinx-immaterial/issues/281#issuecomment-1700933026
tox target to publish on pypi and make gh release https://github.com/jaraco/skeleton/blob/928e9a86d61d3a660948bcba7689f90216cc8243/tox.ini#L42-L58

**Comment by bgoglin** [comment]
Regarding releases, let's discuss in #625
